Purchases: Update cancellation flow copy

drewblaisdell opened the pull request:
This PR addresses part of #2392 by updating `ManagePurchase` and `CancelPurchase` with the copy changes from that issue.

#### Testing instructions

- Run `git checkout update/cancellation-copy` and start your server
- Apply this patch `code-D1469`

*Subscriptions*
- Open `/purchases/:site/:purchaseId` for a non-domain subscription that has auto-renew enabled (e.g. a plan).
- Assert that the link to cancel reads 'Cancel Subscription and Refund', or 'Cancel Subscription' if it is out of the refund window.
- Click the aforementioned cancel link.
- Assert that purchases which are refundable match [this screenshot](https://cloudup.com/c30EJIftlSi).
- Assert that purchases which are not refundable match [this screenshot](https://cloudup.com/cBWvH1teVxX).
- Assert that the refund amount is correct

*Domains*
- Open `/purchases/:site/:purchaseId` for a domain purchase that has auto-renew enabled (e.g. a plan)
- Assert that the link to cancel reads 'Cancel Domain and Refund', or 'Cancel Domain' if it is out of the refund window.
- Click the aforementioned cancel link.
- Assert that domains that are within the refund window match [this screenshot](https://cloudup.com/c0BiX3bKFuT) (the refund amount may differ)
- Assert that domains that are not within the refund window match [this screenshot](https://cloudup.com/c69B4sw0WXB).
- Assert that the refund amount is correct

*One time purchases*
- Open `/purchases/:site/:purchaseId` for a one time purchase (e.g. a premium theme) that was purchased within the last 30 days.
- Assert that the link to cancel reads 'Cancel and Refund'.
- Click the aforementioned cancel link.
- Assert that the cancel page that appears matches [this screenshot](https://cloudup.com/cdghzdenTTD), with different values depending on the theme.
- Assert that the refund amount is correct

#### Additional notes

There are more changes to the flow in #2392 but these will be handled in separate PRs.

#### Reviews

- [x] Code
- [x] Product

stephanethomas commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -68,7 +69,25 @@ const CancelPurchase = React.createClass( {
+		} else {
+			heading = this.translate( 'Cancel Your %(purchaseName)s Subscription', {
+				args: {
+					purchaseName: purchaseName
```
You could use object property shorthands here and above.

stephanethomas commented on client/me/purchases/product-link/index.jsx:
```diff
@@ -0,0 +1,65 @@
+
+		if ( isTheme( selectedPurchase ) ) {
+			url = getThemeDetailsUrl( { id: selectedPurchase.meta }, { slug: selectedPurchase.domain } );
+			text = i18n.translate( 'Theme Details' );
```
I assume you want to make this component stateless at some point. If not, we should probably use `this.translate()` here and above.

stephanethomas commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -63,12 +64,53 @@ const CancelPurchase = React.createClass( {
+	renderFooterText() {
+		const purchase = getPurchase( this.props ),
+			{ priceText, renewDate } = purchase,
+			renewalDate = this.moment( renewDate ).format( 'LL' );
```
Probably not a big deal in term of performance but we could declare this date after the first `if` clause - so it's not instanciated nor formatted for nothing when the subscription is refundable.

stephanethomas commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -7,46 +7,56 @@ import React from 'react';
+				"you'll receive a refund and it will be removed from your site immediately.",
+				{
+					args: {
+						refundPeriodInDays,
```
There is an unnecessary trailing coma here.

fabianapsimoes commented:
In https://github.com/Automattic/wp-calypso/issues/2392#issuecomment-200402575, Spencer and Mike agreed on a few changes to the copy that are not addressed in this PR:

![screenshot-calypso localhost 3000 2016-03-31 15-28-02](https://cloud.githubusercontent.com/assets/1248436/14177147/3acdb55e-f755-11e5-92e0-bc7661f49e80.png)

In this screen, Cancel Subscription After Expiration should only be Cancel Subscription. Then in the following screen...

![screenshot-calypso localhost 3000 2016-03-31 15-33-18](https://cloud.githubusercontent.com/assets/1248436/14177334/160a2166-f756-11e5-8d9a-2e827e980221.png)

- update the button to Cancel Subscription
- update where it says the subscription will "end" to "expire"
- update where it says "to be cancelled" to "will be removed"

I also left a comment in https://github.com/Automattic/wp-calypso/issues/2392#issuecomment-203940765 with regards to the copy for canceling domains. Just giving a heads-up here since it might change.

fabianapsimoes commented:
- Shouldn we update the flow for cancelling Privacy Protection?

- In this branch, I'm getting the following error when trying to open Manage Purchases for legacy upgrades: ```TypeError: inst.render is not a function```. 

- We're showing domain copy for Premium Theme cancellations:

![screenshot-calypso localhost 3000 2016-03-31 16-42-09](https://cloud.githubusercontent.com/assets/1248436/14179503/99b6be6c-f75f-11e5-86b1-e642c95877d8.png)

We shouldn't say "<Theme Name> Subscription" as well, since Premium Themes are one-time purchases. Same for the button.

Also, the link in Theme Details goes to `/theme/:theme/details/:site`. In production, I see that we actually link to `/themes/:theme/:site` which isn't Calypso. If that link will go there in production, we should make sure it opens in a new tab.

fabianapsimoes commented:
> We're showing domain copy for Premium Theme cancellations:

Same for plans within the refund window:

![screenshot-calypso localhost 3000 2016-03-31 16-52-38](https://cloud.githubusercontent.com/assets/1248436/14179878/12a7317a-f761-11e5-93b1-cadcf18abbcf.png)

drewblaisdell commented:
Thanks for the review, @fabianapsimoes and @stephanethomas. I updated the PR to:

- Address the aforementioned copy issues.
- Handle one time purchases – I forgot that these could be canceled within the refund window.
- Fix legacy upgrades – we didn't update the null returns when we converted `ProductLink` into a stateless component.
- Open the theme details link in a new window when `manage/themes/details` is disabled.

> Should we update the flow for cancelling Privacy Protection?

[Here](https://cloudup.com/cAHOeWh8Rh7) is that screen for reference. The language seems straightforward to me in the screenshot, but might be confusing because of the way we use "cancel" elsewhere. Either way, we should update it in a separate PR.

stephanethomas commented:
The `Manage Purchase` and `Cancel Purchase` pages for Google Apps are broken:

![manage purchase](https://cloud.githubusercontent.com/assets/594356/14202310/7e6a6116-f7f7-11e5-8415-57c2211d140a.png)

![cancel purchase](https://cloud.githubusercontent.com/assets/594356/14202312/80d06cde-f7f7-11e5-983d-3ad08f0a67aa.png)

scruffian commented:
Good catch @stephanethomas.

@drewblaisdell I added 0ceffa44a2dc2f17404c615ed80a7ba7f09c0c8d to fix this issue. It seems to me that deciding a purchase is a subscription by checking its not a registration or a one time purchase is easier than listing all the purchase types. If you disagree feel free to revert 0ceffa44a2dc2f17404c615ed80a7ba7f09c0c8d and carefully research every product type we have ever sold :)

fabianapsimoes commented:
- For a Jetpack purchase, View Plan Features shows `/plans/compare` for WordPress.com plans.

- I'm also seeing this in Cancel Purchase:

![screenshot-calypso localhost 3000 2016-04-01 16-26-41](https://cloud.githubusercontent.com/assets/1248436/14209877/a3ecd200-f826-11e5-8a5d-3e4bcb06aa84.png)

scruffian commented:
> For a Jetpack purchase, View Plan Features shows /plans/compare for WordPress.com plans.

Fixed in b774321

> I'm also seeing this in Cancel Purchase:

Fixed in the patch

fabianapsimoes commented:
> Fixed in b774321

Looks good. AFAIK, before this change, we were not showing the PlanAction buttons. Now we are. I don't think it's a big issue, but thought I'd point it out.

Thanks for all the fixes. Product :+1:

stephanethomas commented on client/lib/purchases/index.js:
```diff
@@ -163,6 +167,15 @@ function isRenewing( purchase ) {
 	return includes( [ 'active', 'autoRenewing' ], purchase.expiryStatus );
 }
 
+function isSubscription( purchase ) {
```
It would probably help to add a docblock and explain the concept of subscription. We use this term quite a lot everywhere and here it has a specific meaning, mostly for presentation.

stephanethomas commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -63,12 +64,52 @@ const CancelPurchase = React.createClass( {
+		if ( isRefundable( purchase ) ) {
+			return this.translate( '%(refundText)s to be refunded', {
+				args: { refundText },
+				context: 'refundText is of the form "[currency-symbol][amount]" i.e. "$20"'
```
We could probably use a different name for `refundText` to better convey what it actually contains:

```javascript
renderFooterText() {
		const purchase = getPurchase( this.props ),
			{ refundText: amount, renewDate } = purchase;

		if ( isRefundable( purchase ) ) {
			return this.translate( '%(amount)s to be refunded', {
				args: { amount },
				context: 'amount is of the form "[currency-symbol][amount]" i.e. "$20"'
			} );
		}
```

Tug commented:
LGTM :+1:

stephanethomas commented:
I had two minor comments - nothing that would prevent you from shipping that pull request.
